### PR TITLE
:sparkles: kubectl support ../../ & ..:..:

### PR DESCRIPTION
**Pull request by mjudeikis**
## Summary

Enables to go back more than 1 workspace at the time `ws use ../../`
Enables to use `tree` in rootless clusters. 
```
[mjudeikis@unknown2 kcp]$ k ws tree
.
kcp git:(mjudeikis/refactor.ws.use) k ws tree
.
└── kvdk2spgmbix
    └── test
        └── test1
            └── test2

➜  kcp git:(mjudeikis/refactor.ws.use) k ws use :
Current workspace is ':kvdk2spgmbix'.
➜  kcp git:(mjudeikis/refactor.ws.use) k ws use test:test1:test2
Current workspace is ':kvdk2spgmbix:test:test1:test2'.
➜  kcp git:(mjudeikis/refactor.ws.use) k ws use ../..
Current workspace is ':kvdk2spgmbix:test'.
``` 

Missing parts:
1. `../../foo` multiple actions together does not work
2. no auto-completetion. 

## Release Notes

```release-note
kcp ws use support for relative and absolute multi-step navigation 
```

### CODE REVIEW

**Comment by mjudeikis** [comment]
We need bigger story to refactor all this function to support syntax like:
```
ws use ../../foo
ws use ../../foo/bar
```
basically make it recursive and enable replace `:` with `/` for more linux native navigation,
and add alias to use `ws cd` :)

**Comment by mjudeikis** [comment]
@sttts is this missing smth?

**Comment by embik** [comment]
Overall idea:

```
# absolute path
$ kubectl ws :adct:org-a:team-b
# relative path (called from within adct "root" workspace)
$ kubectl ws org-a:team-b
# relative path (starting from `.` current workspace)
$ kubectl ws .:org-a:team-b
# going from org-a "back" to org-a
$ kubectl ws ..:org-b:..:org-a

# we warn with this case because it's old behaviour
$ kubectl ws root:org-a
warn: did you mean :root:org-a?
```

File: cli/pkg/workspace/plugin/kubeconfig.go [hunk 1]
```diff
@@ -132,187 +134,189 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
...
 		defer func() {
 			if err == nil {
 				_, err = fmt.Fprintf(o.Out, "Note: 'kubectl ws' now matches 'cd' semantics: go to home workspace. 'kubectl ws -' to go back. 'kubectl ws .' to print current workspace.\n")
 			}
 		}()
 		fallthrough
 
-	case "~", home:
-		homeWorkspace, err := o.kcpClusterClient.Cluster(core.RootCluster.Path()).TenancyV1alpha1().Workspaces().Get(ctx, "~", metav1.GetOptions{})
+	case o.Name == "~" || o.Name == home:
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig.go:160]: what about `~:foo` ?

File: cli/pkg/workspace/plugin/kubeconfig.go [hunk 2]
```diff
@@ -132,187 +134,189 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
...
+		// root legacy move
+	case o.Name == core.RootCluster.String():
+		newServerHost, err := o.moveRootLegacy(ctx)
 		if err != nil {
 			return err
 		}
-		u, currentClusterName, err := pluginhelpers.ParseClusterURL(config.Host)
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
+
+	case strings.HasPrefix(o.Name, "system") || strings.HasPrefix(o.Name, ":system"):
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig.go:204]: why is `system:` different from `root:`?

File: cli/pkg/workspace/plugin/kubeconfig_test.go [hunk 3]
```diff
@@ -407,7 +407,7 @@ func TestUse(t *testing.T) {
 				AuthInfos: map[string]*clientcmdapi.AuthInfo{"test": {Token: "test"}},
 			},
 			destination: "root:foo:bar",
-			wantStdout:  []string{"Current workspace is \"root:foo:bar\""},
+			wantStdout:  []string{"Current workspace is ':root:foo:bar'"},
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig_test.go:410]: am not convinced we want to print out the `:` at all. Let's keep the output always `foo:bar:abc`. The `:` prefix is only for navigation.

Comment (mjudeikis) [reply cli/pkg/workspace/plugin/kubeconfig_test.go:410]: was one the line with this, so this pushes swings to same :) will remove

File: cli/pkg/workspace/plugin/kubeconfig.go (hunk 2 above)
Comment (mjudeikis) [reply cli/pkg/workspace/plugin/kubeconfig.go:204]: I think we can remove this and just use `:system:admin` and will just be handled as "absolute any root"

File: cli/pkg/workspace/plugin/kubeconfig.go (hunk 1 above)
Comment (mjudeikis) [reply cli/pkg/workspace/plugin/kubeconfig.go]: added support, just no test as need second run on it :/
~ not being resolved in tests to check if workspace exists gets bit complex in tests :./

File: cli/pkg/workspace/plugin/kubeconfig.go (hunk 2 above)
Comment (mjudeikis) [reply cli/pkg/workspace/plugin/kubeconfig.go]: Added comments. I think it was originally a separate case due to permissions too. Merged it to generic code but it got a complex mess again :/ better have explisting check than if/else/if/else in other places :/

**Comment by mjudeikis** [comment]
reopen

File: cli/pkg/workspace/plugin/kubeconfig.go [hunk 4]
```diff
@@ -132,187 +134,212 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
...
-			fmt.Fprintf(o.ErrOut, "error checking APIBindings: %v\n", err)
-		}
-		if err = findUnresolvedPermissionClaims(o.Out, bindings); err != nil {
-			// display the error, but don't stop the current workspace from being reported.
-			fmt.Fprintf(o.ErrOut, "error checking APIBindings: %v\n", err)
-		}
-
-		return currentWorkspace(o.Out, newServerHost, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
+	// LEGACY(mjudeikis): Remove once everybody gets used to this
+	if strings.HasPrefix(o.Name, "root") {
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig.go:138]: `== "home" || HasPrefix(o.Name, "root:")` ?

File: cli/pkg/workspace/plugin/kubeconfig.go [hunk 5]
```diff
@@ -132,187 +134,212 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
...
 		defer func() {
 			if err == nil {
 				_, err = fmt.Fprintf(o.Out, "Note: 'kubectl ws' now matches 'cd' semantics: go to home workspace. 'kubectl ws -' to go back. 'kubectl ws .' to print current workspace.\n")
 			}
 		}()
 		fallthrough
 
-	case "~", home:
-		homeWorkspace, err := o.kcpClusterClient.Cluster(core.RootCluster.Path()).TenancyV1alpha1().Workspaces().Get(ctx, "~", metav1.GetOptions{})
+	case o.Name == "~" || o.Name == home || strings.HasPrefix(o.Name, "~:") || strings.HasPrefix(o.Name, home):
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig.go:160]:
```suggestion
	case o.Name == "~" || o.Name == home || strings.HasPrefix(o.Name, "~:") || strings.HasPrefix(o.Name, home + ":"):
```

File: cli/pkg/workspace/plugin/kubeconfig.go [hunk 6]
```diff
@@ -132,187 +134,212 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
...
 		}
+		return printCurrentWorkspace(o.Out, newServerHost, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
 
-		// We keep the old host, and append homeworkspace url. This allows users to have front-proxy
-		// host to be different from the workspace host.
-		newServerHost = u.Scheme + "://" + path.Join(u.Host, uh.Path)
-
-	case ".":
-		cfg, err := o.ClientConfig.ClientConfig()
+	case strings.Contains(o.Name, ".."):
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig.go:189]: this one looks fishy. Could see `= ".."`, but anything beyond that needs some more logic, e.g. for `./../foo/../bar`.

File: cli/pkg/workspace/plugin/kubeconfig.go [hunk 7]
```diff
@@ -132,187 +134,212 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
...
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
 
-	default:
-		cluster := logicalcluster.NewPath(o.Name)
-		if !cluster.IsValid() {
-			return fmt.Errorf("invalid workspace name format: %s", o.Name)
-		}
-		config, err := o.ClientConfig.ClientConfig()
+		// root legacy move
+	case o.Name == core.RootCluster.String():
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig.go:197]: could we use `core.RootCluster` everywhere where we refer to `"root"`?

File: cli/pkg/workspace/plugin/kubeconfig.go [hunk 8]
```diff
@@ -132,187 +134,212 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
...
 		if err != nil {
 			return err
 		}
-		u, currentClusterName, err := pluginhelpers.ParseClusterURL(config.Host)
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
+
+		// system is bit different. Usually when moving to system workspaces you
+		// need to have special access and "if workspace exists" checks fails.
+		// If merged to "absolute path check" code gets convoluted :/
+	case strings.HasPrefix(o.Name, ":system"):
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig.go:207]: We have a constant for `system` somewhere I believe.

Also here `HasPrefix` seems fishy. We have to properly split by `":"`.

File: cli/pkg/workspace/plugin/kubeconfig.go (hunk 6 above)
Comment (mjudeikis) [reply cli/pkg/workspace/plugin/kubeconfig.go:189]: yes, this is not supported yet. Code code quite convoluted and wanna move it to separate pr. So this is not yet supported

File: cli/pkg/workspace/plugin/kubeconfig.go (hunk 8 above)
Comment (mjudeikis) [reply cli/pkg/workspace/plugin/kubeconfig.go:207]: Idea was not to trigger this path is somebody created `root:system` or `foo:bar:system` in any ways. So only way to access `system` from cli is to be explicit - `:system`. 
Basically don't want this code path to be triggered by mistake by users.

**Comment by mjudeikis** [comment]
@sttts you ok merge without `../../foo` ?

File: cli/pkg/workspace/plugin/kubeconfig.go (hunk 8 above)
Comment (sttts) [reply cli/pkg/workspace/plugin/kubeconfig.go]: My comment was about using string comparison. Convert it to a logical cluster path and operate on that.

Comment (mjudeikis) [reply cli/pkg/workspace/plugin/kubeconfig.go]: Updated but this does not make it prettier in my opinion :D

File: cli/pkg/workspace/plugin/kubeconfig.go [hunk 9]
```diff
@@ -132,187 +136,210 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
...
-			fmt.Fprintf(o.ErrOut, "error checking APIBindings: %v\n", err)
-		}
-		if err = findUnresolvedPermissionClaims(o.Out, bindings); err != nil {
-			// display the error, but don't stop the current workspace from being reported.
-			fmt.Fprintf(o.ErrOut, "error checking APIBindings: %v\n", err)
-		}
-
-		return currentWorkspace(o.Out, newServerHost, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
+	// LEGACY(mjudeikis): Remove once everybody gets used to this
+	if o.Name == "home" || strings.HasPrefix(o.Name, core.RootCluster.String()+":") {
```
Comment (sttts) [inline cli/pkg/workspace/plugin/kubeconfig.go:140]: `o.Name == home`?

Comment (sttts) [reply cli/pkg/workspace/plugin/kubeconfig.go:140]: Hmm, that case is below. Did we have special "home" logic?

Comment (mjudeikis) [reply cli/pkg/workspace/plugin/kubeconfig.go:140]: we did. I just kept it here. Maybe not needed anymore?

**Pull request merged**